src/sentry/sentry_metrics/indexer/models.py

```diff
@@ -50,3 +58,146 @@ class Meta:
+            organization_id, string = key.split(":")
+            org_ids.append(int(organization_id))
+            strings.append(string)
+
+        db_objects = cls.objects.filter(
+            organization_id__in=org_ids,
+            string__in=strings,
+        )
+        db_results = {}
+        reverse_db_results = {}
+        for db_result in db_objects:
+            key = f"{db_result.organization_id}:{db_result.string}"
+            if key in keys:
+                db_results[key] = db_result.id
+                reverse_key = f"{db_result.organization_id}:{db_result.id}"
+                reverse_db_results[reverse_key] = db_result.string
+
+        indexer_cache.set_many(db_results, field_type=str)
+        indexer_cache.set_many(reverse_db_results, field_type=int)
+        return db_results
+
+    @classmethod
+    def get_id(cls, org_id, string) -> int:
+        key = f"{org_id}:{string}"
+        result = indexer_cache.get(key, field_type=str)
+        if result:
+            metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "true"})
+            return result
+
+        metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "false"})
+        db_result = cls.objects.get(organization_id=org_id, string=string).id
+        indexer_cache.set(key, db_result, field_type=str)
+
+        return db_result
+
+    @classmethod
+    def get_string(cls, org_id, id) -> str:
+        key = f"{org_id}:{id}"
+        result = indexer_cache.get(key, field_type=str)
+        if result:
+            metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "true"})
+            return result
+
+        metrics.incr(_INDEXER_CACHE_METRIC, tags={"cache_hit": "false"})
+        db_result = cls.objects.get(organization_id=org_id, id=id).string
+        indexer_cache.set(key, db_result, field_type=int)
+
+        return db_result
+
+    def delete(self, **kwargs):
+        # TODO(meredith): remove from cache on delete.
+        # maybe should be done with post_delete signal instead
+        return super().delete(**kwargs)
+
+
+class StringIndexerCache:
+    def __init__(self, model, version):
+        self.model = model
+        self.version = version
+        self.cache_ttl = settings.SENTRY_METRICS_INDEXER_CACHE_TTL
+
+    def make_cache_key(self, key: str, field_type: type) -> str:
+        hashed = md5_text(key).hexdigest()
+        if field_type == int:
+            cache_key = f"indexer:org:int:{hashed}"
+        elif field_type == str:
+            cache_key = f"indexer:org:str:{hashed}"
+        else:
+            raise Exception("Invalid type: must be str or int")
+
+        return cache_key
+
+    def get(self, key: str, field_type: type):
+        return cache.get(self.make_cache_key(key, field_type), version=self.version)
+
+    def set(self, key: str, value: Union[str, int], field_type: type):
+        return cache.set(
+            key=self.make_cache_key(key, field_type),
+            value=value,
+            timeout=self.cache_ttl,
```

Comment:
Please introduce a jitter in the cache TTL. This prevents a thundering herd of requests going to the DB when many values expire at the same time.
When we start the system for the first time. A large amount of reads will hit the DB. All those values will be set in cache at the same time. They all have the same TTL so they will all expire at the same time causing a massive a mount of read on the DB. This is called thundering herd.

So what you can do is to add a jitter: take the constant TTL value and add a random value between 0 and 25% of the TTL.

Reply:
Do I need this for `set` I wrote a comment for this under `set_many` since my plan was to use `set_many` when recording a bunch of strings at once. My thought was `set` would only be used if we are looking up an individual string and couldn't find it. I could do the jitter for both though, assuming that I can use `set_many` with different ttls...

Reply:
please do it for both. Of course, when doing it for `set_many` you can apply the same jitter to all the values you are setting. in that same call.